File: sdks/swift/Sources/AstroSDK/Webhooks/WebhookVerifier.swift
import Foundation
import CryptoKit

public struct WebhookVerifier {
    private let secret: String

    public init(secret: String) {
        self.secret = secret
    }

    public func verify(rawBody: String, signature: String) -> Bool {
        let provided = signature.hasPrefix("sha256=") ? String(signature.dropFirst(7)) : signature
        let expected = hmac(rawBody)
        return constantTimeEquals(expected, provided)
    }

    public func parse(rawBody: String) throws -> [String: Any] {
        guard
            let data = rawBody.data(using: .utf8),
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            throw AstroError(statusCode: 0, code: "PARSE_ERROR", message: "Invalid JSON webhook body")
        }
        return json
    }

    public func handle(
        rawBody: String,
        signature: String,
        handlers: [String: ([String: Any]) -> Void]
    ) throws {
        guard verify(rawBody: rawBody, signature: signature) else {
            throw AstroError(statusCode: 401, code: "INVALID_SIGNATURE", message: "Webhook signature verification failed")
        }
        let payload = try parse(rawBody: rawBody)
        guard let event = payload["event"] as? String else { return }
        handlers[event]?(payload)
        handlers["*"]?(payload)
    }

    private func hmac(_ body: String) -> String {
        let keyData = Data(secret.utf8)
        let bodyData = Data(body.utf8)
        let key = SymmetricKey(data: keyData)
        let mac = HMAC<SHA256>.authenticationCode(for: bodyData, using: key)
        return Data(mac).map { String(format: "%02x", $0) }.joined()
    }

    private func constantTimeEquals(_ a: String, _ b: String) -> Bool {
        guard a.count == b.count else { return false }
        let aBytes = Array(a.utf8)
        let bBytes = Array(b.utf8)
        var result: UInt8 = 0
        for i in 0..<aBytes.count { result |= aBytes[i] ^ bBytes[i] }
        return result == 0
    }
}

File: sdks/flutter/lib/src/webhooks/webhook_verifier.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

class WebhookVerifier {
  final String secret;

  const WebhookVerifier(this.secret);

  bool verify(String rawBody, String signature) {
    final provided = signature.startsWith('sha256=')
        ? signature.substring(7)
        : signature;
    final expected = _hmac(rawBody);
    return _constantTimeEquals(expected, provided);
  }

  String _hmac(String body) {
    final key = utf8.encode(secret);
    final bytes = utf8.encode(body);
    final hmac = Hmac(sha256, key);
    return hmac.convert(bytes).toString();
  }

  static bool _constantTimeEquals(String a, String b) {
    if (a.length != b.length) return false;
    var result = 0;
    for (var i = 0; i < a.length; i++) {
      result |= a.codeUnitAt(i) ^ b.codeUnitAt(i);
    }
    return result == 0;
  }

  Map<String, dynamic> parse(String rawBody) =>
      jsonDecode(rawBody) as Map<String, dynamic>;

  void handleRaw(
    String rawBody,
    String signature,
    Map<String, void Function(Map<String, dynamic>)> handlers,
  ) {
    if (!verify(rawBody, signature)) {
      throw Exception('Invalid webhook signature');
    }
    final payload = parse(rawBody);
    final event = payload['event'] as String?;
    if (event == null) return;
    handlers[event]?.call(payload);
    handlers['*']?.call(payload);
  }
}
